packages/date/src/index.js

```diff
@@ -380,34 +383,54 @@ export function date( dateFormat, dateValue = new Date() ) {
  * @param {(Date|string|Moment|null)}        dateValue  Date object or string,
  *                                                      parsable by moment.js.
  *
- * @return {string} Formatted date.
+ * @return {string} Formatted date in English.
  */
 export function gmdate( dateFormat, dateValue = new Date() ) {
 	const dateMoment = momentLib( dateValue ).utc();
 	return format( dateFormat, dateMoment );
 }
 
 /**
- * Formats a date (like `date_i18n()` in PHP).
+ * Formats a date (like `wp_date()` in PHP), translating it into site's locale.
+ *
+ * Backward Compatibility Notice: if `timezone` is set to `true`, the function
+ * behaves like `gmdateI18n`.
+ *
+ * @param {string}                      dateFormat PHP-style formatting string.
+ *                                                 See php.net/date.
+ * @param {(Date|string|Moment|null)}   dateValue  Date object or string,
+ *                                                 parsable by moment.js.
+ * @param {(string|number|null)}        timezone   Timezone to output result in or
+ *                                                 a UTC offset.
+ *                                                 Defaults to timezone from site.
+ *                                                 See momentjs.com.
+ *
+ * @return {string} Formatted date.
+ */
+export function dateI18n( dateFormat, dateValue = new Date(), timezone ) {
```

Comment:
The `timezone` argument isn't consistent with the documented type, in two ways:

- It's optional, but the type doesn't indicate this ([related documentation](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types))
- The first line of code in the function tests if `timezone === true`, but `true` is not a valid documented value for `timezone`. We should include `boolean` if it's expected.

Small formatting nit-picks, but relevant if there's a concern of running out of available space for the description texts:

1. The enclosing parentheses on union types are unnecessary (the `timezone` types as currently implemented could be expressed `{string|number|null}`)
2. You have a few unnecessary extra spaces after your longest type tag `@param {(Date|string|Moment|null)}   dateValue`. There only need be a single space here.

Reply:
I see now a few lines earlier from this that it's mentioned in the documentation about the special treatment for the boolean variable for the sake of backwards-compatibility. One thing I might be worried about is that as we transition to stricter type-checking (#18838), the type-checker may struggle to reconcile or otherwise reject the fact that we document the function as not accepting a boolean, only to then proceed to test a boolean.

I think it may just be worth including `boolean` in the documented type, both for this reason and because it's still technically accepted. Between the documentation of the function (and perhaps an additional note in the argument description), it should be clear enough that `boolean` is effectively deprecated, even if still supported.

Reply:
Fixed in c3d5fe4.

* `timezone` in `dateI18n` now includes `boolean` and explicitly states that "is effectively deprecated, but still supported for backward compatibility reasons."
* Also fixed union types and spacing.

Reply:
There's still some issue here; these functions have optional arguments, but `null` is not a correct way to denote an optional argument (via [reference documentation](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types) referenced in https://github.com/WordPress/gutenberg/pull/18982#discussion_r398132714).